Exposing Kubeflow over load balancer

surajkota opened the pull request:
**Which issue is resolved by this Pull Request:**
Resolves #67 
- Automated script, doc and e2e tests for the feature
- Moved relevant content from Cognito README to load balancer guide
- Reduced the no. of steps for setting up load balancer controller

**Testing:**
- Unit tests: Will create a separate PR. Waiting for #156
- [x] e2e tests pass [test-run.log](https://github.com/awslabs/kubeflow-manifests/files/8449203/test-run.txt)
- added test for lb integration in e2e tests
- Manually tested readme instructions
 
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

goswamig commented on awsconfigs/common/istio-ingress/overlays/https/params.env:
```diff
@@ -0,0 +1 @@
+certArn=
```
Can we add a comment here on purpose of this variable? 
I understand the README.md will have the details but adding a line here doesn't hurt.

goswamig replied:
I would also suggest to create another overlay for http.

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
```diff
@@ -0,0 +1,111 @@
+
+Kubeflow does not offer a generic solution for connecting to Kubeflow over a load balancer because this process is highly dependent on your environment/cloud provider. On AWS, we use the [AWS Load Balancer controller](https://kubernetes-sigs.github.io/aws-load-balancer-controller/) which satisfies the Kubernetes [Ingress resource](https://kubernetes.io/docs/concepts/services-networking/ingress/)  ...
+
+In order to connect to Kubeflow using a LoadBalancer, we need to setup HTTPS. The reason is that many of the Kubeflow web apps (e.g., Tensorboard Web App, Jupyter Web App, Katib UI) use [Secure Cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#restrict_access_to_cookies), so accessing Kubeflow with HTTP over a non-localhost domain does not work.
```
What about other components like pipeline, serving ? Do they also not support HTTP ?

surajkota replied:
looks like kserve webapp too based on: https://www.kubeflow.org/docs/external-add-ons/kserve/webapp/#port-forwarding

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
```diff
@@ -0,0 +1,111 @@
+
+In order to connect to Kubeflow using a LoadBalancer, we need to setup HTTPS. The reason is that many of the Kubeflow web apps (e.g., Tensorboard Web App, Jupyter Web App, Katib UI) use [Secure Cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#restrict_access_to_cookies), so accessing Kubeflow with HTTP over a non-localhost domain does not work.
+
+To secure the traffic and use HTTPS, we must associate a Secure Sockets Layer/Transport Layer Security (SSL/TLS) certificate with the load balancer. [AWS Certificate Manager](https://aws.amazon.com/certificate-manager/) is a service that lets you easily provision, manage, and deploy public and private Secure Sockets Layer/Transport Layer Security (SSL/TLS) certificates for use with AWS services a ...
```
Can you link the reference doc that mentioned this limitation ?

`certificates cannot be created for ALB DNS`

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
```diff
@@ -0,0 +1,111 @@
+After installing the components, the script will wait for the ingress managed load balancer to be created and update the subdomain (`platform.example.com`) with a CNAME entry for `kubeflow.platform.example.com` pointing to the load balancer DNS. You can see the load balancer DNS by running `kubectl get ingress -n istio-system istio-ingress`.
+
+To summarize, what the script does or what you need to if you are doing it manually:
+    - Create a custom domain to host Kubeflow
```
`domain and subdomain`

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
```diff
@@ -0,0 +1,111 @@
+
+To summarize, what the script does or what you need to if you are doing it manually:
+    - Create a custom domain to host Kubeflow
+    - Create TLS certificates for the domain using ACM
```
`domain and subdomain`

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
```diff
@@ -0,0 +1,111 @@
+    - Create TLS certificates for the domain using ACM
+    - Configure the ingress and load balancer controller
+    - Build manifests and install the required components
+    - Update the domain with ALB address
```
`subdomain`

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
```diff
@@ -0,0 +1,111 @@
+
+As explained in the background section, you need a registered domain and TLS certificate to use HTTPS with load balancer. Since your top level domain(e.g. `example.com`) could have been registered at any service provider, for uniformity and taking advantage of the integration provided between Route53, ACM and Application Load Balancer, you will create a separate [sudomain](https://en.wikipedia.or ...
+
+### Ingress and Load Balancer Controller configuration
```
If I am assuming it right, you need to write precise command that user has to run by following this guide.

surajkota replied:
Addressed this comment in the latest version of guide. Previously, I was trying to only keep automated script the primary way to do this.

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
````diff
@@ -0,0 +1,111 @@
+
+1. Install dependencies for the script
+    ```
+    cd tests/e2e
````
why we are asking to moving to tests directory if its user facing README?

surajkota replied:
This is a limitation we are facing with using python right now to run these as part of tests, validate script dependencies and avoid any code duplication. We took this trade off so the scripts can be run in same way in tests or if user runs it

We can investigate how to enhance this but it is out of scope for this PR.

surajkota replied:
https://github.com/awslabs/kubeflow-manifests/issues/166

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
````diff
@@ -0,0 +1,111 @@
+    cd tests/e2e
+    pip install -r requirements.txt
+    ```
+1. Substitute values in `tests/e2e/utils/load_balancer/config.yaml`.
````
I am lost here, are we writing this steps for developer to test this feature ? or we writing this for end user ?

surajkota replied:
answered above

surajkota commented on docs/deployment/cognito/README.md:
```diff
@@ -92,16 +89,20 @@ In this section, we will be creating certificate to enable TLS authentication at
-            2. ![cognito-domain-cloudfront-A-record-creating](./images/cognito-domain-cloudfront-A-record-creatin.png)
-            3. Following is a screenshot of the A record for `auth.platform.example.com` in `platform.example.com` hosted zone:
-                1. ![cognito-domain-cloudfront-A-record-created](./images/cognito-domain-cloudfront-A-record-created.png)
+1. Add a custom domain to the user pool. In order to add a custom domain to your user pool, you need specify a domain name, and provide a certificate managed with AWS Certificate Manager (ACM).
```
I was trying to make the domain and cert creation generic to reuse it across different guides and hence moved the cognito specific things. I think the new structure in this revision will make it clear

surajkota commented on docs/deployment/cognito/README.md:
```diff
@@ -13,7 +13,7 @@ This guide assumes that you have:
     - [jq](https://stedolan.github.io/jq/download/) - A command line tool for processing JSON.
     - [kustomize version 3.2.0](https://github.com/kubernetes-sigs/kustomize/releases/tag/v3.2.0) - A command line tool to customize Kubernetes objects through a kustomization file.
-      - :warning: Kubeflow 1.3.0 is not compatible with the latest versions of of kustomize 4.x. This is due to changes in the order resources are sorted and printed. Please see [kubernetes-sigs/kustomize#3794](https://github.com/kubernetes-sigs/kustomize/issues/3794) and [kubeflow/manifests#1797](https://github.com/kubeflow/manifests/issues/1797). We know this is not ideal and are working with t ...
+      - :warning: Kubeflow is not compatible with the latest versions of of kustomize 4.x. This is due to changes in the order resources are sorted and printed. Please see [kubernetes-sigs/kustomize#3794](https://github.com/kubernetes-sigs/kustomize/issues/3794) and [kubeflow/manifests#1797](https://github.com/kubeflow/manifests/issues/1797). We know this is not ideal and are working with the ups ...
```
Will do in future

surajkota commented on tests/e2e/utils/load_balancer/lb_resources_cleanup.py:
```diff
@@ -0,0 +1,120 @@
+                # load balancer controller does not place a finalizer on the ingress and so deleting the attached load balancer is asynhronous
+                # adding a random wait to allow controller to delete the load balancer
+                # TODO: implement a better check
+                time.sleep(2 * 60)
```
Added a better check with exception if deletion does not happen

surajkota commented:
@goswamig I have made significant changes to the guide from the last review and addressed comments as well. Please take a look

@ryansteakley can you take a look as well?

goswamig commented:
I would suggest to use 

also there are some error, specifically if we are not deleting the resources, over the time they will pilled up.

```
ingress.networking.k8s.io "istio-ingress" deleted
INFO:__main__:d7ee0b0f-istiosystem-istio-2af2-1382882308.ca-central-1.elb.amazonaws.com: waiting for deletion ...
ERROR:e2e.utils.aws.elbv2:Failed to describe load balancer d7ee0b0f-istiosystem-istio-2af2
Traceback (most recent call last):
  File "/home/ubuntu/kubeflow/kubeflow-manifests/tests/e2e/utils/aws/elbv2.py", line 28, in describe
    response = self.elbv2_client.describe_load_balancers(
  File "/home/ubuntu/anaconda3/envs/kf/lib/python3.8/site-packages/botocore/client.py", line 388, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/ubuntu/anaconda3/envs/kf/lib/python3.8/site-packages/botocore/client.py", line 708, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.LoadBalancerNotFoundException: An error occurred (LoadBalancerNotFound) when calling the DescribeLoadBalancers operation: Load balancers '[d7ee0b0f-istiosystem-istio-2af2]' not found
2022-04-08 06:27:46 [ℹ]  eksctl version 0.82.0
2022-04-08 06:27:46 [ℹ]  using region ca-central-1
2022-04-08 06:27:46 [ℹ]  1 iamserviceaccount (kubeflow/alb-ingress-controller) was included (based on the include/exclude rules)
2022-04-08 06:27:46 [ℹ]  1 task: { 
    2 sequential sub-tasks: { 
        delete IAM role for serviceaccount "kubeflow/alb-ingress-controller" [async],
        delete serviceaccount "kubeflow/alb-ingress-controller",
    } }2022-04-08 06:27:46 [ℹ]  will delete stack "eksctl-surakota-dev-10-addon-iamserviceaccount-kubeflow-alb-ingress-controller"
2022-04-08 06:27:46 [ℹ]  deleted serviceaccount "kubeflow/alb-ingress-controller"
ERROR:e2e.utils.aws.iam:failed to delete iam policy arn:aws:iam::169544399729:policy/alb_ingress_controller_surakota-dev-10gyche6w5po
Traceback (most recent call last):
  File "/home/ubuntu/kubeflow/kubeflow-manifests/tests/e2e/utils/aws/iam.py", line 51, in delete
    self.iam_client.delete_policy(
  File "/home/ubuntu/anaconda3/envs/kf/lib/python3.8/site-packages/botocore/client.py", line 388, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/ubuntu/anaconda3/envs/kf/lib/python3.8/site-packages/botocore/client.py", line 708, in _make_api_call
    raise error_class(parsed_response, operation_name)

```

surajkota commented:
@goswamig The two errors you see:
1. failed to describe alb - script waits after deleting ingress for the ALB to be deleted. The error log appears when ALB is successfully deleted
2. for IAM policy error - there is also because of a retry. There is a delay in IAM service for the attached role deletion to reflect in the policy. It fails initially after the role is deleted but as soon as policy is up to date, the deletion is successful
```
INFO:e2e.utils.aws.iam:deleted iam policy arn:aws:iam::169544399729:policy/alb_ingress_controller_surakota-dev-10gyche6w5po
```

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
```diff
@@ -0,0 +1,192 @@
+
+## Create Load Balancer
+
+To make it easy to create the load balancer, you can use the [script provided in this section](#automated-script). If you prefer to use rhe automated scripts, you need to only execute the steps in the [automated script section](#automated-script). Read the following sections in this guide to understand what happens when you run the script or execute all the steps if you prefer to do it manually/h ...
```
typo:  If you prefer to use `rhe` automated scripts

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
````diff
@@ -0,0 +1,192 @@
+                    aws ec2 create-tags --resources ${i} --tags Key=kubernetes.io/cluster/${CLUSTER_NAME},Value=${TAG_VALUE}
+                done
+                ```
+        1. `kubernetes.io/role/internal-elb`. Add this tag only to private subnet
+        1. `kubernetes.io/role/elb`. Add this tag only to public subnet.
````
I think there are multiple private and public subnet, right ?
if so,
subnet -> subnets

goswamig commented on docs/deployment/add-ons/load-balancer/README.md:
````diff
@@ -0,0 +1,192 @@
+    1. ```
+        kubectl get ingress -n istio-system istio-ingress
+        NAME            CLASS    HOSTS   ADDRESS                                                                  PORTS   AGE
+        istio-ingress   <none>   *       ebde55ee-istiosystem-istio-2af2-1100502020.us-west-2.elb.amazonaws.com   80      15d
````
would suggest put dummy address.

surajkota replied:
this is dummy address, replace few chars with xxx to make it more explicit

goswamig commented on tests/e2e/utils/load_balancer/setup_load_balancer.py:
```diff
@@ -0,0 +1,262 @@
+
+
+def create_ingress():
+    kubectl_apply_kustomize(path=common.LB_KUSTOMIZE_PATH)
```
where do i see this file ? the only file in path `LB_KUSTOMIZE_PATH = "../../docs/deployment/add-ons/load-balancer/"` is README.md am I missing something here?

surajkota replied:
https://github.com/surajkota/kubeflow-manifests/blob/lb_dex/docs/deployment/add-ons/load-balancer/kustomization.yaml


its part of this PR, can you check again?

goswamig commented on tests/e2e/utils/load_balancer/setup_load_balancer.py:
```diff
@@ -0,0 +1,262 @@
+    region: str, cluster_name: str, subdomain_hosted_zone: Route53HostedZone
+) -> str:
+
+    wait_for_alb_dns(cluster_name, region)
+    ingress = get_ingress(cluster_name, region)
+    alb_dns = ingress["status"]["loadBalancer"]["ingress"][0]["hostname"]
+    wait_for_alb_status(alb_dns, region)
```
makes sense, the dns gets populated first before marking the status active.

goswamig reviewed (commented):
almost there,  few minor comments and questions.
